feat(proxy): accounts endpoint group

NunoAlexandre opened the pull request:
Closes https://github.com/radicle-dev/radicle-upstream/issues/667.

Note: This PR is branched off https://github.com/radicle-dev/radicle-upstream/pull/675 for now, which is waiting on a registry PR to be merged.

MeBrei commented on proxy/src/http/account.rs:
```diff
@@ -0,0 +1,245 @@
+        )))
+        .and(document::document(
+            document::body(http::RegisterProjectInput::document()).mime("application/json"),
+        ))
```
`RegisterProjectInput` not needed here

MeBrei commented on proxy/src/http/account.rs:
```diff
@@ -0,0 +1,245 @@
+            .await;
+
+        http::test::assert_response(&res, StatusCode::OK, |have| {
+            assert_eq!(have, json!(0), "Account was expected to exist on chain");
```
```suggestion
            assert_eq!(have, json!(0), "Account was not expected to exist on chain");
```

NunoAlexandre commented:
> Looks good so far. I inlined 2 small comments.
> 
> I tried to test both requests in the API playground. The balance request returned `Response Status: Failed to execute 'fetch' on 'Window': Request with GET/HEAD method cannot have body. (CORS or Network Issue) ` and the account exists request said that the account did not exist, though I used the account id returned from the fetch user request.

My bad, I am sorry for the poor quality check. Thanks for your thorough look.

One thing more to discuss: Currently, if an account does **not** exist on chain, its balance will be 0. Do we want to return an error or a 404 in such case? We could introduce that error at the proxy registry side and if that error occurs on the api side we return a 404?

NunoAlexandre commented on proxy/src/http.rs:
```diff
@@ -101,6 +103,7 @@ where
 
     let api = path("v1").and(combine!(
         avatar_filter,
+        account_filter,
```
For reviewers: I added this here instead of before `avatar_filter` because doing the latter failed to compile. Since we will merge `avatar_filter` into `ids`, I just put it here for now.

xla replied:
Addressed in 8953b2f2a1466b2871e2810d128ff7926c73e88a

MeBrei commented:
> One thing more to discuss: Currently, if an account does **not** exist on chain, its balance will be 0. Do we want to return an error or a 404 in such case? We could introduce that error at the proxy registry side and if that error occurs on the api side we return a 404?

I think we should return a 404. It is quite unlikely that it would happen with the current use cases (ie we only pass the account id we get for the user/org from registry) but especially then, we should make an error noticeable in the ui

NunoAlexandre commented:
> > One thing more to discuss: Currently, if an account does **not** exist on chain, its balance will be 0. Do we want to return an error or a 404 in such case? We could introduce that error at the proxy registry side and if that error occurs on the api side we return a 404?
> 
> I think we should return a 404. It is quite unlikely that it would happen with the current use cases (ie we only pass the account id we get for the user/org from registry) but especially then, we should make an error noticeable in the ui

Done :v:

xla commented on proxy/src/http/account.rs:
```diff
@@ -0,0 +1,225 @@
+where
+    R: registry::Client,
+{
+    http::with_shared(registry)
```
We usually put the dependencies after the verb, path, etc.

MeBrei commented on proxy/src/http.rs:
```diff
@@ -66,6 +67,7 @@ where
     let store = Arc::new(RwLock::new(store));
     let subscriptions = crate::notification::Subscriptions::default();
 
+    let account_filter = path("accounts").and(account::filters(&Arc::clone(&registry)));
```
I would keep `path("accounts")` in `account.rs` so it is consistent with the other declarations of the path

NunoAlexandre replied:
That was @xla. I would prefer that too.

xla replied:
We gonna move all of the other ones out. It's prefactoring, the great promise of a golden future to come.

NunoAlexandre replied:
Ok. That wold cut the duplication that we currently have setting filters for testing purposes :ok_hand:

xla commented on proxy/src/http/id.rs:
```diff
@@ -17,7 +17,7 @@ where
         .and(http::with_shared(registry))
         .and(warp::get())
-        .and(document::param::<registry::Id>(
+        .and(document::param::<String>(
```
Would be great if we had `try_param`. :>

NunoAlexandre replied:
@xla not sure what you are referring to :) could you elaborate? To avoid misunderstandings it's helpful to read this commit's description in case you haven't yet: https://github.com/radicle-dev/radicle-upstream/pull/681/commits/12529a5b31741cfd52b1f8226ddcb69d62cc080f

xla replied:
My comment was in regards that `param` doesn't seem to support types which only implement `TryFrom`, which would be handy in this instance.

NunoAlexandre replied:
ah I see! that's right, will follow up on that. As the diff shows, I moved this one back to `String` to have the docs playground working but I will check whether using `try_param::<registry::Id>` solves the issue. :+1:

NunoAlexandre replied:
@xla, `document::try_param` is not a thing apparently?

xla replied:
That's why I said it would be great if it existed. :dancers:

NunoAlexandre replied:
[papoy](https://youtu.be/l0tOJ3gyokk?t=9)! I thought you were proposing to change it :b

NunoAlexandre commented:
Before we merge this, I want to fix the issues we are seeing regarding the interaction with this new endpoint.

More specifically, I have two issues I want to address.

### Issue 1 - Interacting with the endpoints - Good and Bad requests

Merle realized that playing with it via the docs playground didn't work, even though it works via the Rust unit tests calling the endpoint. I noticed that the `ids/<>/status` endpoint didn't work from the docs playground either, despite it working from the tests, ui, and the terminal. I fixed the ids case with https://github.com/radicle-dev/radicle-upstream/pull/681/commits/12529a5b31741cfd52b1f8226ddcb69d62cc080f.

Now, for these new `accounts` endpoint, I see it working correctly when I use a well-formed account id:

- exists

```
 ➜ curl http://127.0.0.1:8080/v1/accounts/5FA9nQDVg267DEd8m1ZypXLBnvN7SFxYwV7ndqSYGiN9TTpu/exists     
true%
```

- balance

```
 ➜ curl http://127.0.0.1:8080/v1/accounts/5FA9nQDVg267DEd8m1ZypXLBnvN7SFxYwV7ndqSYGiN9TTpu/balance     
1152921504606846976%
```


**However**, if we provide an ill-formed account id, it fails:

```
 ➜ curl http://127.0.0.1:8080/v1/accounts/foo/exists                                                  
{"message":"Resource not found","variant":"NOT_FOUND"}%
```

The respective error code being `500`:

```
➜ curl -I http://127.0.0.1:8080/v1/accounts/foo/exists                                               
HTTP/1.1 500 Internal Server Error
```


This is no good. For ill-formed input, the returned status code should be `400 Bad Request` and should contain a message describing what's wrong.

**How** I _think_ we could fix this by expecting the param to be a String and deserialize it within the handler. If it fails to deserialize, we then return a 400 explaining it received an invalid account id.


### Issue 2 - Getting `undefined` when automating testing these endpoints from the ui

See `ui/src/account.test.ts`. Both the tests fails with:

```
Expected: <value>
Received: undefined
```

Note they fail using well-formed input that works on the terminal. For a moment I thought I had forgotten to mock the api but that's done. Any idea? :thinking:

NunoAlexandre commented:
@MeBrei Alright, so I have dropped `account.test.ts` since it was pointless to test these calls against a mocked api. I have fixed the issues reported in issue #1, just waiting for the companion branch in the registry to be merged.

@xla It's weird to me that to fail with `404` one does `reject::not_found` but to fail with `400` I need to `Ok(reply...`. I couldn't find a way of return an Err to return a 400, which would help dry and clean the code around this a little. Do you have any tips? We can also address that in a follow up.
